instance/models/server.py

```diff
@@ -242,18 +252,36 @@ def _set_progress(self, progress, expected_status=None):
         # The '_progress' field needs to be saved:
         self.save()
 
-    def sleep_until_status(self, *target_status_list):
+    def sleep_until(self, condition):
         """
-        Sleep in a loop until the server reaches one of the specified statuses
+        Sleep in a loop until condition related to server status is fulfilled.
+
+        Raises an exception if the desired condition can not be fulfilled.
+        This can happen if the server is in a steady state (i.e., a state that is not expected to change)
+        that does not fulfill the desired condition.
+
+        Use as follows:
+
+            server.sleep_until(lambda: server.status.is_steady_state)
+            server.sleep_until(lambda: server.status.accepts_ssh_commands)
         """
-        self.logger.info('Waiting to reach status %s...', target_status_list)
+        if condition():
+            return
+        else:
+            if self.status.is_steady_state:
+                raise SteadyStateException(
+                    "The current state ({name}) does not fulfill the desired condition "
+                    "and is steady, i.e., it is not expected to change.".format(name=self.status.name)
+                )
+
+        self.logger.info('Waiting to reach status from which we can proceed...')
 
         while True:
             self.update_status()
-            if self.progress.is_final and self.status.one_of(*target_status_list):
+            if self.progress.is_final and condition():
+                self.logger.info('Reached appropriate status. Proceeding.')
```

Comment:
BTW, this isn't part of this task, but if you have time, I'd love to see us add a timeout parameter to this method as well, so that if the loop sleeps for more than, say, a default of 60 minutes, it will raise an exception.